src/externalized/strategic/UndergroundSectorModel.cc

```diff
@@ -0,0 +1,126 @@
+#include "UndergroundSectorModel.h"
+
+#include "Random.h"
+
+UndergroundSectorModel::UndergroundSectorModel(uint8_t sectorId_, uint8_t sectorZ_, uint8_t adjacentSectors_, 
+	std::array<uint8_t, NUM_DIF_LEVELS> numTroops_, std::array<uint8_t, NUM_DIF_LEVELS> numElites_, std::array<uint8_t, NUM_DIF_LEVELS> numCreatures_, 
+	std::array<uint8_t, NUM_DIF_LEVELS> numTroopsVariance_, std::array<uint8_t, NUM_DIF_LEVELS> numElitesVariance_, std::array<uint8_t, NUM_DIF_LEVELS> numCreaturesVariance_) 
+		: sectorId(sectorId_), sectorZ(sectorZ_), adjacentSectors(adjacentSectors_),
+		numTroops(numTroops_), numElites(numElites_), numCreatures(numCreatures_),
+		numTroopsVariance(numTroopsVariance_), numElitesVariance(numElitesVariance_), numCreaturesVariance(numCreaturesVariance_) {}
+
+UNDERGROUND_SECTORINFO* UndergroundSectorModel::createUndergroundSectorInfo(uint8_t difficultyLevel) const
+{
+	UNDERGROUND_SECTORINFO* const u = new UNDERGROUND_SECTORINFO{};
+
+	u->ubSectorX = SECTORX(sectorId);
+	u->ubSectorY = SECTORY(sectorId);
+	u->ubSectorZ = sectorZ;
+	u->ubAdjacentSectors = adjacentSectors;
+
+	u->ubNumTroops    = numTroops[difficultyLevel - 1]    + Random(numTroopsVariance[difficultyLevel - 1]   );
+	u->ubNumElites    = numElites[difficultyLevel - 1]    + Random(numElitesVariance[difficultyLevel - 1]   );
+	u->ubNumCreatures = numCreatures[difficultyLevel - 1] + Random(numCreaturesVariance[difficultyLevel - 1]);
+
+	return u;
+}
+
+std::array<uint8_t, NUM_DIF_LEVELS> readIntArray(const rapidjson::Value& obj, const char* fieldName)
+{
+	std::array<uint8_t, NUM_DIF_LEVELS> vals = {};
+	if (!obj.HasMember(fieldName))
+	{
+		return vals;
+	}
+	auto arr = obj[fieldName].GetArray();
+	if (arr.Size() != NUM_DIF_LEVELS)
+	{
+		SLOGE(ST::format("The number of values in {} is not same as NUM_DIF_LEVELS({})", fieldName, NUM_DIF_LEVELS));
+		throw std::runtime_error("The number of elements does not match the number of difficulty levels");
```

Comment:
Do we catch these exceptions somewhere or are they fatal? It seems two logging calls are redundant.

Reply:
They are unhandled exceptions. On Windows it will be a message box saying "exited with status code -107374091". Cannot see the error message without attaching a debugger. Also, I have not figuredd  out how to properly do `ST::String` and subtitutions with std::runtime_error.

Need a bit guidance and suggestions here. I can add this to #1072 to clean up all occurences.

Reply:
It appears unification would only be possible for debug builds (SLOGA). Are you saying the native message box will not contain our message or were you referring to the title?

Reply:
Yes I am saying the message box will not contain our message, so the user cannot tell what went wrong.

Reply:
So then isn't it simplest to just throw with no message, since it's getting logged beforehand anyway?

Reply:
Updated as said.